service/mapper/transitions.go

```diff
@@ -160,50 +158,54 @@ func (t *Transitions) BootstrapState(s *State) error {
 
 			// save registers in batch, which could result better speed
 			if len(batch) >= batchSize {
-				nBatch.Inc()
 				jobs <- batch
 				batch = make([]*wal.LeafNode, 0, batchSize)
 			}
 		}
 
 		if len(batch) > 0 {
-			nBatch.Inc()
 			jobs <- batch
 		}
-		close(jobs)
 	}()
 
-	results := make(chan error)
+	// wait for all workers to finish in order to close the results channel
+	var wg sync.WaitGroup
+	wg.Add(nWorker)
+
+	// create enough buffer to hold the max number of errors,
+	// there will be at most nWorker number of errors, because a worker
+	// will stop as soon as hitting an error
+	workerErrors := make(chan error, nWorker)
+	// create nWorker number of workers to process jobs.
+	// each worker will save a batch of registers to storage
 	for w := 0; w < nWorker; w++ {
 		go func() {
-			left := int32(1)
+			defer wg.Done()
+
+			// process batches from jobs channel
 			for batch := range jobs {
 				err := t.write.Registers(s.height, batch)
-				results <- err
+
 				if err != nil {
+					workerErrors <- err
 					cancel()
 					return
 				}
 
-				left = nBatch.Dec()
 				select {
 				case <-ctx.Done():
 					return
 				default:
 				}
 			}
-
-			// no more jobs
-			// TODO: fix error handling
-			// only one worker will finish the last nBatch
-			if left == 0 {
-				log.Info().Msgf("finished processing all payloads")
-				close(results)
-			}
 		}()
 	}
 
-	for err := range results {
+	wg.Wait()
+	close(workerErrors)
+
+	// read from results channel and check for errors
+	for err := range workerErrors {
```

Comment:
would it make sense to start reading from the channel immediately and abort the workers on the first error?

Reply:
Yeah, the thing is, `range workerErrors` would cause deadlock unless `close` is called, that's why I need to use the wait group to wait for all workers to finish and close the `workerErrors` channel, before iterating the channel with `workerErrors`. 

In other words, if I just start reading from the workerErrors channel, in the happy case, where there is no error, then program will run into deadlock. 

When a worker running into some error, cancel will be called, and it will stop reading jobs, and all workers will stop soon as there is no more jobs. Yes, it's possible that there might be multiple errors from different workers, and we only return one of the errors, but I guess that's OK, because this is the bootstrap phase, if there is any error, we will crash and have to check and redo everything all over again anyway.